jsonschema/validators.py

```diff
@@ -56,24 +61,91 @@ def _validates(cls):
     return _validates
 
 
-def create(meta_schema, validators=(), version=None, default_types=None):
+def _generate_legacy_type_checks(types=()):
+    """
+    Generate type check definitions suitable for TypeChecker.redefine_many,
+    using the supplied types. Type Checks are simple isinstance checks,
+    except checking that numbers aren't really bools.
+
+    Arguments:
+
+        types (dict):
+
+            A mapping of type names to their Python Types
+
+    Returns:
+
+        A dictionary of definitions to pass to TypeChecker
+
+    """
+    types = dict(types)
+
+    def gen_type_check(pytypes):
+        pytypes = _utils.flatten(pytypes)
+
+        def type_check(checker, instance):
+            if isinstance(instance, bool):
+                if bool not in pytypes:
+                    return False
+            return isinstance(instance, pytypes)
+
+        return type_check
+
+    definitions = {}
+    for typename, pytypes in iteritems(types):
+        definitions[typename] = gen_type_check(pytypes)
+
+    return definitions
+
+class ValidatorMetaClass(type):
+    @property
+    def DEFAULT_TYPES(self):
+        warn("DEFAULT_TYPES is deprecated, use type_checker",
+             DeprecationWarning)
+        return self._DEFAULT_TYPES
+
+
+def create(meta_schema, validators=(), version=None, default_types=None,
+           type_checker=None):
+
+    use_default_types=False
+    if default_types is not None or type_checker is None:
+        use_default_types = True
+        warn("default_types is deprecated, use type_checker",
+             DeprecationWarning)
+
     if default_types is None:
         default_types = {
             u"array": list, u"boolean": bool, u"integer": int_types,
             u"null": type(None), u"number": numbers.Number, u"object": dict,
             u"string": str_types,
         }
 
+    if type_checker is None:
+        type_checker = _types.TypeChecker()
+
+    if use_default_types:
+        type_checker = type_checker.redefine_many(
+            _generate_legacy_type_checks(default_types))
+
+    @add_metaclass(ValidatorMetaClass)
     class Validator(object):
         VALIDATORS = dict(validators)
         META_SCHEMA = dict(meta_schema)
-        DEFAULT_TYPES = dict(default_types)
+        _DEFAULT_TYPES = dict(default_types)
+        TYPE_CHECKER = type_checker
 
         def __init__(
-            self, schema, types=(), resolver=None, format_checker=None,
-        ):
-            self._types = dict(self.DEFAULT_TYPES)
-            self._types.update(types)
+            self, schema, types=(), resolver=None, format_checker=None):
+
+            if types:
```

Comment:
This won't warn in the (unlikely but possible) case that someone provides a falsy value to types, even though that code will break when the argument is removed.

Reply:
Good catch. It should check for anything other than an empty tuple.